fugue/execution/native_execution_engine.py

```diff
@@ -279,6 +280,35 @@ def sample(
             ValueError("one and only one of n and frac should be set"),
         )
         d = df.as_pandas().sample(n=n, frac=frac, replace=replace, random_state=seed)
+        return PandasDataFrame(d.reset_index(drop=True), df.schema, metadata)
+
+    def limit(
+        self,
+        df: DataFrame,
+        n: int,
+        presort: str,
+        partition_spec: PartitionSpec = EMPTY_PARTITION_SPEC,
+        metadata: Any = None,
+    ) -> DataFrame:
+        assert_or_throw(
+            isinstance(n, int),
+            ValueError("n needs to be an integer"),
+        )
+        d = df.as_pandas()
+
+        if presort:
+            presort = _parse_presort_exp(presort)
+        # Use presort over partition_spec.presort if possible
+        _presort: IndexedOrderedDict = presort or partition_spec.presort
+
+        if len(_presort.keys()) > 0:
+            d = d.sort_values(list(_presort.keys()), ascending=list(_presort.values()))
+
+        if len(partition_spec.partition_by) == 0:
+            d = d.head(n)
+        else:
+            d = d.groupby(by=partition_spec.partition_by, dropna=False).head(n)
+
```

Comment:
I tried to use pl_utils.safe_groupby_apply but I got errors so I did this for now and it passes the tests.